### PR TITLE
Add validator FSM support

### PR DESCRIPTION
**Pull request by flotothemoon**
Adds support for the recently introduced validator registration FSM from Radix engine. Specifically, the API is expanded with two new actions: `RegisterValidatorAction` and `UnregisterValidatorAction` with the corresponding mappers. To reduce logic duplication, validator registration state mapping is implemented in the shared `ValidatorRegistrationState` helper.

### CODE REVIEW

File: examples/src/main/java/com/radixdlt/client/examples/ValidatorExample.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.radixdlt.client.examples;
```
Comment (clstrfsck) [inline examples/src/main/java/com/radixdlt/client/examples/ValidatorExample.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java [hunk 2]
```diff
@@ -774,6 +780,38 @@ public Result sendTokens(
...
+	/**
+	 * Registers the given address as a validator.
+	 *
+	 * @param validator the validator address to be registered
+	 * @return result of the transaction
+	 */
+	public Result registerValidator(
+		RadixAddress validator
+	) {
+		Objects.requireNonNull(validator, "validator");
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java]: Would it be better to do this null check in the public constructor for RegisterValidatorAction?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java [hunk 3]
```diff
@@ -774,6 +780,38 @@ public Result sendTokens(
...
+	/**
+	 * Unregisters the given address as a validator.
+	 *
+	 * @param validator the validator address to be unregistered
+	 * @return result of the transaction
+	 */
+	public Result unregisterValidator(
+		RadixAddress validator
+	) {
+		Objects.requireNonNull(validator, "validator");
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java]: Similarly here, would it be better to do this null check in the public constructor for UnregisterValidatorAction?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * The registration state of a certain validator address.
+ */
+final class ValidatorRegistrationState {
+	private final transient Particle particle;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java]: Not sure why this is marked "transient" here?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		UnregisteredValidatorParticle initialState = new UnregisteredValidatorParticle(address, 0);
+		return ValidatorRegistrationState.from(initialState);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java]: Would prefer instanceof check for final class.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	 * @return the initial {@link ValidatorRegistrationState} for the address
+	 */
+	static ValidatorRegistrationState initial(RadixAddress address) {
+		Objects.requireNonNull(address, "address");
+		UnregisteredValidatorParticle initialState = new UnregisteredValidatorParticle(address, 0);
+		return ValidatorRegistrationState.from(initialState);
+	}
+
+	@Override
+	public boolean equals(Object o) {
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:131]: Would like to see an EqualsVerifier unit test for this.

File: radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		// for serializer
+	}
+
+	public UnregisteredValidatorParticle(RadixAddress address, long nonce) {
+		super(address.euid());
+		this.address = address;
+		this.nonce = nonce;
+	}
+
+	public RadixAddress getAddress() {
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java:58]: Should have @ Override from Ownable.

File: radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+		// for serializer
+	}
+
+	public RegisteredValidatorParticle(RadixAddress address, long nonce) {
+		super(address.euid());
+		this.address = address;
+		this.nonce = nonce;
+	}
+
+	public RadixAddress getAddress() {
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java:58]: Should have @ Override from Ownable.

File: radixdlt-java/src/test/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationStateTest.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/test/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationStateTest.java:23]: Missing header comments.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorAction.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorAction.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorAction.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorAction.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
+package com.radixdlt.client.application.translate.validators;
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:23]: Missing header comment.

File: radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java:1]: I think this is the wrong license header for the library.  Should be the MIT header.

File: radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (clstrfsck) [inline radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java:1]: This is the wrong header comment for the library.  Should be the MIT header.

**Review by clstrfsck — CHANGES_REQUESTED** [review]
A couple of missing header comments and some other items that could be considered.

File: radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java (hunk 15 above)
Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java:1]: Indeed, will replace.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java (hunk 4 above)
Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java]: Brainfart. Removing.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java (hunk 2 above)
Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java]: That would be nicer, yeah.

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Action mapper for {@link RegisterValidatorAction}s, implemented using {@link ValidatorRegistrationState}.
+ */
+public class RegisterValidatorActionMapper implements StatefulActionToParticleGroupsMapper<RegisterValidatorAction> {
+	@Override
+	public Set<ShardedParticleStateId> requiredState(RegisterValidatorAction action) {
+		return ImmutableSet.of(
+			ShardedParticleStateId.of(RegisteredValidatorParticle.class, action.getValidator()),
+			ShardedParticleStateId.of(UnregisteredValidatorParticle.class, action.getValidator())
```
Comment (talekhinezh) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java:48]: I would think you only need UnregisteredValidatorParticle here no?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Action mapper for {@link UnregisterValidatorAction}s, implemented using {@link ValidatorRegistrationState}.
+ */
+public class UnregisterValidatorActionMapper implements StatefulActionToParticleGroupsMapper<UnregisterValidatorAction> {
+	@Override
+	public Set<ShardedParticleStateId> requiredState(UnregisterValidatorAction action) {
+		return ImmutableSet.of(
+			ShardedParticleStateId.of(RegisteredValidatorParticle.class, action.getValidator()),
+			ShardedParticleStateId.of(UnregisteredValidatorParticle.class, action.getValidator())
```
Comment (talekhinezh) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java:48]: I would think you only need RegisteredValidatorParticle here no?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+import com.radixdlt.identifiers.RadixAddress;
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * The registration state of a certain validator address.
+ */
+final class ValidatorRegistrationState {
```
Comment (talekhinezh) [inline radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:37]: Can we rename this as its easy to confuse with it being an ApplicationState

Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:37]: True, though it being local to the validators package makes that somewhat less worrisome. Happy to rename anyway, do you have suggestions?

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java (hunk 17 above)
Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java:48]: Leaving for now as discussed

File: radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java (hunk 18 above)
Comment (flotothemoon) [reply radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java:48]: Leaving for now as discussed

**Review by clstrfsck — APPROVED** [review]

**Pull request merged**